Code/Framework/AzCore/AzCore/base.h

```diff
@@ -15,9 +15,19 @@
 
 #ifndef AZ_ARRAY_SIZE
 /// Return an array size for static arrays.
-#   define  AZ_ARRAY_SIZE(__a)  (sizeof(__a)/sizeof(__a[0]))
+namespace AZ::Internal
+{
+        template<class T>
+        struct StaticArraySize
+        {
+            static_assert(std::is_array_v<T>, "AZ_ARRAY_SIZE can only be used with a C-style array");
+            static constexpr size_t value = sizeof(T) / sizeof(std::remove_extent_t<T>);
```

Comment:
`std::` is being used here instead of `AZStd` to avoid a circular include dependency in the type traits.